### DISCUSSION TITLE
Analysis of imports and exports with raw untransformed AST

### DISCUSSION THREAD

**Discussion by tmpfs** [swc-project/swc#1776] [General]
I want to analyze the `imports` and `exports` exposed by `TransformedModule` but I also want the underlying module not to be transformed so I can access the original AST.

I think the options are:

1) Load the module using a `Parser` and reproduce the work to analyze the `imports` and `exports`
2) Expose an API on `Bundler` that would allow bypassing the AST transformations

My preference would be for 2) - do you think we could achieve this easily?

**Comment by kdy1** [comment]
It's not hard, but I think bundler should not expose such apis.

**Reply by tmpfs** [reply]
In that case would it be possible to move the logic that builds `Imports` and `Exports` to another package so it is re-usable outside the context of `swc_bundler`?

**Reply by kdy1** [reply]
Yes, it's possible

**Reply by tmpfs** [reply]
Great, I would like to try to do this. Would putting the code in `swc_dep_graph` make more sense than creating a new package?

**Reply by kdy1** [reply]
I thought they should go into `/ecmascript/loader`, but I found a more fundamental problem.
The behavior of the loader implementation of bundler is too opinionated for public apis.
It modifies span hygiene data in a way that can be handled by the bundler afterward.
It's an implementation detail and can be changed anytime if I found a better way.

**Reply by tmpfs** [reply]
I looked into the [imports/mod.rs](https://github.com/swc-project/swc/blob/master/bundler/src/bundler/import/mod.rs) and it appears to be be tightly coupled with the `Bundler` so it does not look like it could easily be extracted to another package :(
